### ISSUE TITLE
environment-to-ini tracking bug

### ISSUE THREAD

**Issue by n9** [go-gitea/gitea#27541]
### Description

There are known issues with `environment-to-ini`: https://gitea.com/gitea/helm-chart/issues/356

However, I have not found an issue this repo, where the implementation of `environment-to-ini` is located.

### Gitea Version

1.20.5

### Can you reproduce the bug on the Gitea demo site?

No

### How are you running Gitea?

Currently, the issue only occurs when `environment-to-ini` is explicitly used.

However, @silverwind mentioned in https://github.com/go-gitea/gitea/issues/24804#issuecomment-1554378613 that there is an intention to merge `enviornment-to-ini` into the codebase. (I am not sure, whether this may cause then also issues in other deployments.)

### Database

None

### Further notes

In my case, the issue occurs during the configuration of "custom logger modes":

> Failed to load writer mode "router" for logger router: invalid log writer type (mode): router, maybe it needs something like 'MODE=file' in [log.router] section
> Failed to load writer mode "xorm" for logger xorm: invalid log writer type (mode): xorm, maybe it needs something like 'MODE=file' in [log.xorm] section

This is because `environment-to-ini` is unable to write `[log.router] MODE=console` to `app.ini`, as it is a default value.

(In this case, however, I am not sure what should be fixed, whether the  `environment-to-ini` or "custom logger modes".)

**Comment by n9** [comment]
@wxiaoguang This is not about deletes, but is about "adds".

**Comment by n9** [comment]
Try to run environment-to-ini with e.g.

```
GITEA__log_0X2E_router__MODE=console
GITEA__log_0X2E_xorm__MODE=console
```

These values will not be added to `app.ini`.

**Comment by n9** [comment]
I have added a failing test to demonstrate the issue: https://github.com/n9/gitea/commit/c70896126ac82c4c94eaaf7a3a2a2f76b1a18513

**Comment by wxiaoguang** [comment]
It's the legacy/unmaintained INI package's bug (or quirk behavior).

The section values are inherited. So `[sec].key=SOME` also means  `[sec.sub].key=SOME`

**Comment by n9** [comment]
Yes, but this does not work here in `loadLogModeByName`.

**Comment by wxiaoguang** [comment]
Thank you for your detailed report.

I think the fix could be this:  Fix environment-to-ini inherited key bug #27543

**Comment by n9** [comment]
Thank you for the fix.
